dotnet/src/Connectors/Connectors.Onnx/PoolingMode.cs

```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.SemanticKernel.Connectors.Onnx;
+
+/// <summary>Pooling mode used for creating the final sentence embedding.</summary>
+public enum EmbeddingPoolingMode
+{
+    /// <summary>Uses the maximum across all token embeddings.</summary>
+    Max,
+    /// <summary>Calculates the average across all token embeddings.</summary>
+    Mean,
+    /// <summary>Calculates the average across all token embeddings, divided by the square root of the number of tokens.</summary>
+    MeanSquareRootTokensLength,
```

Comment:
If the goal was to maximize perf, we could have a `Sum` mode that simply skips the final division. It won't make any difference to cosine similarity.

However without benchmarks I don't have much of a sense of whether it would even have a measurable impact.

Reply:
BTW what's SK's position on benchmarking in general? I don't see any in the repo. Are there some somewhere else?

Reply:
> If the goal was to maximize perf

This isn't about perf. The sentence-transformers library defines several pooling strategies for taking the N produced embeddings and consolidating them down to the single fixed-length result. These are three of them. I didn't implement the others yet (eg cls, last_token, weighted). The pooling folder on HF may contain the spec for which should be used, eg
https://huggingface.co/TaylorAI/bge-micro-v2/blob/main/1_Pooling/config.json#L4

Reply:
Sorry if the phrasing was unclear. I don't mean "if pooling is about perf", I just mean "if we want to maximize perf".

Just suggesting that a custom pooling operator ("sum") could skip the final division and produce vectors that are still exactly equivalent under cosine similarity.

I appreciate that we don't know up front if the developer is actually going to use cosine similarity or another metric, so unless the final division has an appreciable cost then I agree it would make sense not to try to avoid it.

Reply:
Ah, gotcha, I misunderstood the comment.  Yeah, I'd prefer to wait to make it configurable until there's a need.

Though I'm actually missing the divide here to make this an actual mean (if I'm reading it right, the smartcomponents code is also missing it); I have that fixed locally, but in doing that I also changed the validation to not use cosine similarity and instead just compare each value with a tolerance threshold, and that highlighted that some models are then employing an additional normalization step, and I'm trying to figure out where that's defined in the HF configuration. Regardless, I think I'll need to add that as an additional option, in order to get the outputs to match what you'd get on the Python side by default.

Reply:
Yeah I see now it is missing the final divide here.

In the Smart Components case that was intentional because it was defined as sum pooling (given that the developer has no choice over the similarity metric).

In the `BertOnnxTextEmbeddingGenerationService` case, where you do control the similarity metric, there would definitely be a stronger case for making the number match up with Python output.